## Q1. Pearson correlation coefficient is a measure of the linear relationship between two variables. Suppose you have collected data on the amount of time students spend studying for an exam and their final exam scores. Calculate the Pearson correlation coefficient between these two variables and interpret the result.

The Pearson correlation coefficient, also known as Pearson's r, is a statistical measure that quantifies the linear relationship between two continuous variables. It indicates both the direction (positive or negative) and the strength of the linear relationship between the variables. The Pearson correlation coefficient ranges from -1 to +1, where -1 represents a perfect negative linear correlation, +1 represents a perfect positive linear correlation, and 0 represents no linear correlation between the variables.

In [1]:
import pandas as pd 
df = pd.DataFrame({ "hours spent" : [12,10,9,14,15,14,6,7,10,5],
              "scores" : [89,84,83,94,97,90,78,82,85,71]
             })
df

hours spent  scores
0           12      89
1           10      84
2            9      83
3           14      94
4           15      97
5           14      90
6            6      78
7            7      82
8           10      85
9            5      71

In [2]:
df.corr(method='pearson')

hours spent    scores
hours spent     1.000000  0.960941
scores          0.960941  1.000000

## Q2. Spearman's rank correlation is a measure of the monotonic relationship between two variables. Suppose you have collected data on the amount of sleep individuals get each night and their overall job satisfaction level on a scale of 1 to 10. Calculate the Spearman's rank correlation between these two variables and interpret the result.

Spearman's rank correlation coefficient, often referred to as Spearman's rho (ρ), is a non-parametric measure of the monotonic relationship between two variables. It assesses the strength and direction of the association between two variables by ranking their values and then computing the Pearson correlation coefficient on the ranked data.

Pearson's correlation, which measures the linear relationship, wheras Spearman's rank correlation captures any monotonic relationship between the variables. A monotonic relationship implies that as the values of one variable increase, the values of the other variable either increase or decrease, but not necessarily at a constant rate.

Spearman's rank correlation coefficient ranges from -1 to +1, where -1 indicates a perfect negative monotonic relationship, +1 indicates a perfect positive monotonic relationship, and 0 suggests no monotonic relationship between the variables. It is commonly used when the data does not meet the assumptions of normality or when the relationship between variables is non-linear.

In [3]:
import pandas as pd

amount_of_sleep = [7, 6, 8, 5, 6, 7, 8, 6, 5, 7]
job_satisfaction = [8, 7, 9, 3, 4, 7, 10, 7, 6, 8]

df = pd.DataFrame({"sleep hour": amount_of_sleep, "job satisfaction" : job_satisfaction})
df.head()

sleep hour  job satisfaction
0           7                 8
1           6                 7
2           8                 9
3           5                 3
4           6                 4

In [4]:
df.corr(method= 'spearman' )

sleep hour  job satisfaction
sleep hour          1.000000          0.914401
job satisfaction    0.914401          1.000000

Here we observe from the spearman correlation being positive close to 1 symbolizes the data is highly positvely correlated ie.with the number of sleep hours the overall job satisfaction level increases.

## Q3. Suppose you are conducting a study to examine the relationship between the number of hours of exercise per week and body mass index (BMI) in a sample of adults. You collected data on both variables for 50 participants. Calculate the Pearson correlation coefficient and the Spearman's rank correlation between these two variables and compare the results.

In [5]:
import numpy as np
import pandas as pd
import random

random.seed(42) # for reproducibility

# Generate random data for the number of hours of exercise per week (between 1 and 20 hours)
exercise_hours = np.random.randint(1, 20,size=50)

# Generate random data for body mass index (BMI) (between 15 and 35)
bmi_values = np.random.uniform(15, 35,size=50)

df = pd.DataFrame({"Excercise hours per week " : exercise_hours,
                   "BMI" : bmi_values
                  })
df.head()

Excercise hours per week         BMI
0                          8  18.997367
1                         19  21.070184
2                         19  34.798209
3                          6  15.111290
4                         19  19.130947

In [6]:
df.corr(method='pearson')

Excercise hours per week        BMI
Excercise hours per week                    1.000000  0.049798
BMI                                         0.049798  1.000000

In [7]:
df.corr(method = 'spearman')

Excercise hours per week        BMI
Excercise hours per week                    1.000000  0.057145
BMI                                         0.057145  1.000000

- Based on the above values obtained we see there is a very little difference between the two correlation methods the values coming 0.132 and 0.140 for pearson and spearman respectively.

- Based on this we see there is little or no relationship between the the variables as the values of the coefficient are close to 0.

## Q4. A researcher is interested in examining the relationship between the number of hours individuals spend watching television per day and their level of physical activity. The researcher collected data on both variables from a sample of 50 participants. Calculate the Pearson correlation coefficient between these two variables.

In [8]:
import numpy as np
import pandas as pd
import random

random.seed(382)#for reproduciblity

# Generate random data for the number of hours of physical activity during day (between 1 and 18 hours)
physical_activity_hours = np.random.randint(1,18,size=50)

# Generate random data for the number of hours of watching television during day (between 1 and 18 hours)
watching_televison_hours = np.random.randint(1, 18,size=50)

df = pd.DataFrame({"Physical Activity hours per day " : physical_activity_hours,
                   "Watching Televison hours per day" : watching_televison_hours
                  })
df.head()

Physical Activity hours per day   Watching Televison hours per day
0                                 8                                 6
1                                 5                                 1
2                                 8                                15
3                                 3                                 5
4                                13                                 4

In [9]:
df.corr(method='pearson')

Physical Activity hours per day   \
Physical Activity hours per day                           1.000000   
Watching Televison hours per day                          0.381421   

                                  Watching Televison hours per day  
Physical Activity hours per day                           0.381421  
Watching Televison hours per day                          1.000000

In [10]:
df.corr(method='spearman')

Physical Activity hours per day   \
Physical Activity hours per day                           1.000000   
Watching Televison hours per day                          0.336731   

                                  Watching Televison hours per day  
Physical Activity hours per day                           0.336731  
Watching Televison hours per day                          1.000000

There is no significant diiference between the two methods of correction they are almost neutral as they are negative but close to 0.

## Q5. A survey was conducted to examine the relationship between age and preference for a particular brand of soft drink. The survey results are shown below:

In [11]:
df = pd.DataFrame({'Age(in Years)' : [25,42,37,19,31,28],
                   'Soft Drink Preference' : ["Coke","Pepsi","Mountain Dew","Coke","Pepsi","Coke"]
                  })
df

Age(in Years) Soft Drink Preference
0             25                  Coke
1             42                 Pepsi
2             37          Mountain Dew
3             19                  Coke
4             31                 Pepsi
5             28                  Coke

In [12]:
#First we will convert the soft drink preference to numerical for applying correlation
#We will use OHE for converting
from sklearn.preprocessing import OneHotEncoder

encode = OneHotEncoder()
values = encode.fit_transform(df[['Soft Drink Preference']]).toarray()

encode_df = pd.DataFrame(values,columns=encode.get_feature_names_out())
df = pd.concat([df,encode_df],axis = 1)
df.drop(columns="Soft Drink Preference",inplace=True)
df

Age(in Years)  Soft Drink Preference_Coke  \
0             25                         1.0   
1             42                         0.0   
2             37                         0.0   
3             19                         1.0   
4             31                         0.0   
5             28                         1.0   

   Soft Drink Preference_Mountain Dew  Soft Drink Preference_Pepsi  
0                                 0.0                          0.0  
1                                 0.0                          1.0  
2                                 1.0                          0.0  
3                                 0.0                          0.0  
4                                 0.0                          1.0  
5                                 0.0                          0.0

In [13]:
df.corr(method="pearson")

Age(in Years)  Soft Drink Preference_Coke  \
Age(in Years)                            1.000000                   -0.837240   
Soft Drink Preference_Coke              -0.837240                    1.000000   
Soft Drink Preference_Mountain Dew       0.394132                   -0.447214   
Soft Drink Preference_Pepsi              0.576439                   -0.707107   

                                    Soft Drink Preference_Mountain Dew  \
Age(in Years)                                                 0.394132   
Soft Drink Preference_Coke                                   -0.447214   
Soft Drink Preference_Mountain Dew                            1.000000   
Soft Drink Preference_Pepsi                                  -0.316228   

                                    Soft Drink Preference_Pepsi  
Age(in Years)                                          0.576439  
Soft Drink Preference_Coke                            -0.707107  
Soft Drink Preference_Mountain Dew                    -0.316228  
Soft Drink Preference_Pepsi                            1.000000

In [14]:
df.corr(method="spearman")

Age(in Years)  Soft Drink Preference_Coke  \
Age(in Years)                            1.000000                   -0.878310   
Soft Drink Preference_Coke              -0.878310                    1.000000   
Soft Drink Preference_Mountain Dew       0.392792                   -0.447214   
Soft Drink Preference_Pepsi              0.621059                   -0.707107   

                                    Soft Drink Preference_Mountain Dew  \
Age(in Years)                                                 0.392792   
Soft Drink Preference_Coke                                   -0.447214   
Soft Drink Preference_Mountain Dew                            1.000000   
Soft Drink Preference_Pepsi                                  -0.316228   

                                    Soft Drink Preference_Pepsi  
Age(in Years)                                          0.621059  
Soft Drink Preference_Coke                            -0.707107  
Soft Drink Preference_Mountain Dew                    -0.316228  
Soft Drink Preference_Pepsi                            1.000000

From this we observe :

- People with higher age tend to like Pepsi the most as its the highest positively correlated by a value of 0.62
- People with higher age tend to possibly like Mountain dew but its less positively correlated comapred to pepsi with a value of 0.39
- As the age increases the like of prefrence of choosing coke reduces significantly as it has a correlation value of -0.87 very close to -1 indicating a very strong negative correlation.

## Q6. A company is interested in examining the relationship between the number of sales calls made per day and the number of sales made per week. The company collected data on both variables from a sample of 30 sales representatives. Calculate the Pearson correlation coefficient between these two variables.

In [15]:
import numpy as np
import pandas as pd
import random

# Generate synthetic data
np.random.seed(4323)  # for reproducibility


sales_calls = np.random.randint(20, 40, size=30)
sales_per_week = np.random.uniform(2000, 4000, size=30)
df = pd.DataFrame({"Sales Calls" :sales_calls,
                  "Weekly Sales" : sales_per_week})
df.head()

Sales Calls  Weekly Sales
0           36   3234.082997
1           37   3657.311542
2           26   2014.340867
3           29   2495.656869
4           38   2223.092926

In [16]:
#Calculating Pearson by using Dataframe
df.corr(method='pearson')

Sales Calls  Weekly Sales
Sales Calls      1.000000     -0.101161
Weekly Sales    -0.101161      1.000000

In [17]:
#Calculating Pearson by using Scipy Library
from scipy.stats import pearsonr

pearson_corr , p_value = pearsonr(sales_calls,sales_per_week)

print(f"Pearson Correlation Coefficient : {pearson_corr}")
print(f"P-value : {p_value} ")

Pearson Correlation Coefficient : -0.10116108915369336
P-value : 0.5947910499968548 


In the context of calculating the Pearson correlation coefficient, the null hypothesis typically assumes that there is no significant linear relationship between the two variables being analyzed. In this case, with a p-value of 0.59, it means that there is a 59% chance of observing the correlation coefficient (or a more extreme value) between the number of sales calls made per day and the number of sales made per week, even if there is no true linear relationship between these variables.

Since the p-value is greater than the common significance level (alpha), which is typically set at 0.05, 0.01, or 0.001, depending on the study's requirements, we do not have enough evidence to reject the null hypothesis. In other words, we do not have sufficient evidence to conclude that there is a significant linear relationship between the number of sales calls and the number of sales made per week in the sample data.